compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -1502,6 +1502,9 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                                 pluralize!(count),
                             ),
                         );
+                        if sp.is_desugaring(DesugaringKind::Async) {
+                            err.note("while checking the return type of this `async fn`");
```

Comment:
I'm not sure how cumbersome it might be to procure, but if we had the `Span` for `make_u32` we could have the following output:

```
error[E0308]: mismatched types
  --> $DIR/dont-suggest-missing-await.rs:14:18
   |
LL | async fn make_u32() -> u32 {
   |          --------      --- checked the `Output` of this `async fn`, found opaque type
   |          |
   |          while checking the return type of this `async fn`
...
LL |         take_u32(x)
   |                  ^ expected `u32`, found opaque type
   |
   = note:     expected type `u32`
           found opaque type `impl Future`
help: consider `await`ing on the `Future`
   |
LL |         take_u32(x.await)
   |                   ^^^^^^
```

If getting it is too difficult, then we should reword the `note` to be "while checking the return type of the `async fn`", because the note has no span and the output on both the CLI and third-party tools might be ambiguous.